### PR TITLE
Refactor streaming from client to server

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/aspnet/SignalR/issues/2592
Fixes https://github.com/aspnet/AspNetCore/issues/4401

TODO:
- [x] Couple more tests
- [x] Update spec with new behavior

### CODE REVIEW

**Comment by davidfowl** [comment]
I think we should also support accepting a channel writer for sever to client streaming. This is something we discussed briefly but didn’t go into details on. It’s worth looking at as it could simplify the programming model.

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 1]
```diff
@@ -465,14 +466,26 @@ async Task OnStreamCanceled(InvocationRequest irq)
                         readers = new Dictionary<string, object>();
                     }
 
-                    var id = _connectionState.GetNextStreamId();
+                    var id = _connectionState.GetNextId();
                     readers[id] = args[i];
-                    args[i] = new StreamPlaceholder(id);
+
+                    // TODO: super ugly
+                    Array.Resize(ref args, args.Length - 1);
```
[the remark this answers is not in the document]
Comment (halter73) [reply src/SignalR/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: You could even keep the out parameter as a `List<string>`.

Doesn't `Array.Resize(ref args, args.Length - 1)` remove the last element of the list? And doesn't the `i--` after this make it so you just loop over the same argument repeatedly until you reach the last one?

File: src/SignalR/clients/ts/signalr/src/IHubProtocol.ts [hunk 2]
```diff
@@ -40,9 +36,7 @@ export type HubMessage =
     CompletionMessage |
     CancelInvocationMessage |
     PingMessage |
-    CloseMessage |
-    StreamCompleteMessage |
-    StreamDataMessage;
+    CloseMessage;
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/IHubProtocol.ts]: Should InvocationMessage and StreamInvocationMessage also be combined?

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs [hunk 3]
```diff
@@ -102,6 +132,22 @@ public StreamInvocationMessage(string invocationId, string target, object[] argu
...
+        /// <param name="invocationId">The invocation ID.</param>
+        /// <param name="target">The target method name.</param>
+        /// <param name="arguments">The target method arguments.</param>
+        /// <param name="streams">The target methods stream IDs.</param>
+        public StreamInvocationMessage(string invocationId, string target, object[] arguments, string[] streams)
+            : base(invocationId, target, arguments, streams)
+        {
+            if (string.IsNullOrEmpty(invocationId))
+            {
+                throw new ArgumentNullException(nameof(invocationId));
```
Comment (halter73) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs]: Is this argument check the only difference between StreamInvocationMessage and HubMethodInvocationMessage?

Comment (halter73) [reply src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs]: It seems that the presence of a streams/streamIds property should be enough to distinguish the two types of messages.

Comment (JamesNK) [reply src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs]: Also, throwing an `ArgumentNullException` for an empty string isn't right.

Either throw an `ArgumentException` for both, or `ArgumentException` when it is empty.

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ public class NewtonsoftJsonHubProtocol : IHubProtocol
         private const string ResultPropertyName = "result";
         private const string ItemPropertyName = "item";
         private const string InvocationIdPropertyName = "invocationId";
-        private const string StreamIdPropertyName = "streamId";
+        private const string StreamsPropertyName = "streams";
```
Comment (mikaelm12) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs]: "streamIds"

**Review by mikaelm12 — COMMENTED** [review]
Still reviewing. I think some of my comments are already out of date with the commit you just made seconds ago

File: src/SignalR/clients/ts/signalr/src/IHubProtocol.ts (hunk 2 above)
Comment (halter73) [reply src/SignalR/clients/ts/signalr/src/IHubProtocol.ts]: Couldn't that be determined based on the signature of the hub method being invoked? Or the presence of the stream/streamIds property in the InvocationMessage?

Comment (BrennanConroy) [reply src/SignalR/clients/ts/signalr/src/IHubProtocol.ts]: Current design of the streamIds property is that it is only for streams from client to server.

And we can't rely on the signature of the Hub method to determine if the client intended to wait for a normal invocation response (via CompletionMessage) or a streamed response (via StreaItems + CompletionMessage). So we use two different invocation messages, StreamInvocationMessage and InvocationMessage

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs [hunk 5]
```diff
@@ -112,7 +112,7 @@ public override string ToString()
 
             try
             {
-                streamIds = string.Join(", ", StreamIds?.Select(id => id?.ToString()));
+                streamIds = string.Join(", ", StreamIds != null ? StreamIds.Select(id => id?.ToString()) : Array.Empty<string>());
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs]: Should an error be thrown for a null streamId?

This place might not be the right place but I think it should be guarded against.

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubDispatcher.cs [hunk 6]
```diff
@@ -208,6 +208,13 @@ private Task ProcessStreamItem(HubConnectionContext connection, StreamItemMessag
 
                 try
                 {
+                    var clientStreamLength = hubMethodInvocationMessage.StreamIds != null ? hubMethodInvocationMessage.StreamIds.Length : 0;
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubDispatcher.cs]: nit: hubMethodInvocationMessage.StreamIds?.Length ?? 0

File: src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackHubProtocol.ts [hunk 7]
```diff
@@ -217,31 +216,43 @@ export class MessagePackHubProtocol implements IHubProtocol {
...
 
         return BinaryMessageFormat.write(payload.slice());
     }
 
-    private writeStreamComplete(streamCompleteMessage: StreamCompleteMessage): ArrayBuffer {
+    private writeCompletion(completionMessage: CompletionMessage): ArrayBuffer {
         const msgpack = msgpack5();
-        const payload = msgpack.encode([MessageType.StreamComplete, streamCompleteMessage.streamId,
-            streamCompleteMessage.error || null]);
+        const resultKind = completionMessage.error ? 1 : completionMessage.result ? 3 : 2;
```
Comment (JamesNK) [inline src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackHubProtocol.ts]: Are these values being used in multiple places? For example, both for writing the completionMessage and reading it? Might be worth making them constants.

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.MessagePack/Protocol/MessagePackHubProtocol.cs [hunk 8]
```diff
@@ -289,6 +284,23 @@ private static StreamCompleteMessage CreateStreamCompleteMessage(byte[] input, r
             }
         }
 
+        private static string[] ReadStreamIds(byte[] input, ref int offset)
+        {
+            var streamIdLength = ReadArrayLength(input, ref offset, "streams");
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.MessagePack/Protocol/MessagePackHubProtocol.cs]: I think "streams" here should be "streamIds"

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.MessagePack/Protocol/MessagePackHubProtocol.cs [hunk 9]
```diff
@@ -475,6 +479,22 @@ private void WriteArgument(object argument, Stream stream)
             }
         }
 
+        private void WriteStreamIds(string[] streams, Stream packer)
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.MessagePack/Protocol/MessagePackHubProtocol.cs]: nit: streamIds

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs [hunk 10]
```diff
@@ -131,6 +130,7 @@ private HubMessage ParseMessage(Utf8BufferTextReader textReader, IInvocationBind
                 JToken resultToken = null;
                 var hasArguments = false;
                 object[] arguments = null;
+                string[] streams = null;
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs]: nit streamIds

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs [hunk 11]
```diff
@@ -599,6 +567,22 @@ private void WriteArguments(object[] arguments, JsonTextWriter writer)
             writer.WriteEndArray();
         }
 
+        private void WriteStreamIds(string[] streams, JsonTextWriter writer)
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs]: nit: streamIds

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs [hunk 12]
```diff
@@ -24,7 +24,7 @@ public class NewtonsoftJsonHubProtocol : IHubProtocol
         private const string ResultPropertyName = "result";
         private const string ItemPropertyName = "item";
         private const string InvocationIdPropertyName = "invocationId";
-        private const string StreamIdPropertyName = "streamId";
+        private const string StreamsPropertyName = "streamIds";
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs]: StreamIdsPropertyName

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs [hunk 13]
```diff
@@ -684,7 +643,7 @@ private HubMessage BindStreamItemMessage(string invocationId, object item, bool
             return new StreamItemMessage(invocationId, item);
         }
 
-        private HubMessage BindStreamInvocationMessage(string invocationId, string target, object[] arguments, bool hasArguments, IInvocationBinder binder)
+        private HubMessage BindStreamInvocationMessage(string invocationId, string target, object[] arguments, bool hasArguments, string[] streams, IInvocationBinder binder)
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs]: nit: streamIds

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs [hunk 14]
```diff
@@ -701,10 +660,10 @@ private HubMessage BindStreamInvocationMessage(string invocationId, string targe
                 throw new InvalidDataException($"Missing required property '{TargetPropertyName}'.");
             }
 
-            return new StreamInvocationMessage(invocationId, target, arguments);
+            return new StreamInvocationMessage(invocationId, target, arguments, streams);
         }
 
-        private HubMessage BindInvocationMessage(string invocationId, string target, object[] arguments, bool hasArguments, IInvocationBinder binder)
+        private HubMessage BindInvocationMessage(string invocationId, string target, object[] arguments, bool hasArguments, string[] streams, IInvocationBinder binder)
```
Comment (JamesNK) [inline src/SignalR/src/Microsoft.AspNetCore.SignalR.Protocols.NewtonsoftJson/Protocol/NewtonsoftJsonHubProtocol.cs]: nit: streamIds

File: src/SignalR/test/Microsoft.AspNetCore.SignalR.Tests.Utils/TestClient.cs [hunk 15]
```diff
@@ -95,9 +95,14 @@ public TestClient(IHubProtocol protocol = null, IInvocationBinder invocationBind
...
         }
 
-        public async Task<IList<HubMessage>> StreamAsync(string methodName, params object[] args)
+        public Task<IList<HubMessage>> StreamAsync(string methodName, params object[] args)
         {
-            var invocationId = await SendStreamInvocationAsync(methodName, args);
+            return StreamAsync(methodName, streams: null, args);
+        }
+
+        public async Task<IList<HubMessage>> StreamAsync(string methodName, string[] streams, params object[] args)
```
Comment (JamesNK) [inline src/SignalR/test/Microsoft.AspNetCore.SignalR.Tests.Utils/TestClient.cs]: nit: streamIds

File: src/SignalR/test/Microsoft.AspNetCore.SignalR.Tests.Utils/TestClient.cs [hunk 16]
```diff
@@ -173,14 +178,19 @@ public Task<string> SendInvocationAsync(string methodName, bool nonBlocking, par
         }
 
         public Task<string> SendStreamInvocationAsync(string methodName, params object[] args)
+        {
+            return SendStreamInvocationAsync(methodName, streams: null, args);
+        }
+
+        public Task<string> SendStreamInvocationAsync(string methodName, string[] streams, params object[] args)
```
Comment (JamesNK) [inline src/SignalR/test/Microsoft.AspNetCore.SignalR.Tests.Utils/TestClient.cs]: nit: streamIds

File: src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs (hunk 5 above)
Comment (BrennanConroy) [reply src/SignalR/src/Microsoft.AspNetCore.SignalR.Common/Protocol/HubMethodInvocationMessage.cs]: I don't think we should care if stream ids is null, it's similar to how we don't care about extra data in the protocol

**Review by JamesNK — APPROVED** [review]
:shipit:

File: src/SignalR/test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/TestHubMessageEqualityComparer.cs [hunk 17]
```diff
@@ -158,6 +129,34 @@ private bool SequenceEqual(object left, object right)
...
+            {
+                return false;
+            }
+
+            if (left.Length != right.Length)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < left.Length; i++)
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply src/SignalR/test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/TestHubMessageEqualityComparer.cs]: Add a test with multiple items as well

File: src/SignalR/specs/HubProtocol.md [hunk 18]
```diff
@@ -133,6 +133,10 @@ Ping messages do not have any payload, they are completely empty messages (aside
 
 The default ASP.NET Core implementation automatically pings both directions on active connections. These pings are at regular intervals, and allow detection of unexpected disconnects (for example, unplugging a server). If the client detects that the server has stopped pinging, the client will close the connection, and vice versa. If there's other traffic through the connection, keep-alive pings aren't needed. A `Ping` is only sent if the interval has elapsed without a message being sent.
 
+## Upload streaming
+
+Sometimes Callees will want to receive streaming data from Callers. The Caller can begin such a process by making an `Invocation` or `StreamInvocation` and adding a "StreamIds" property with an array of IDs. The IDs must be unique from any other stream IDs used by the same Caller. The Caller then sends `StreamItem` messages with the "InvocationId" property set to the ID for the stream they are sending over. The Caller ends the stream by sending a `Completion` message with the ID of the stream they are completing.
```
Comment (mikaelm12) [inline src/SignalR/specs/HubProtocol.md]: The Callees doesn't have to initiate this though, right? I think this should be worded more like, The Caller has the ability to to send streaming data to the callee as well , or something.

Comment (mikaelm12) [reply src/SignalR/specs/HubProtocol.md]: nit: for the line reading "and adding a "StreamIds" property with an array of IDs." Add something to the effect of  -> representing the 1 or more streams that will be associated with this invocation

File: src/SignalR/specs/HubProtocol.md [hunk 19]
```diff
@@ -133,6 +133,10 @@ Ping messages do not have any payload, they are completely empty messages (aside
 
 The default ASP.NET Core implementation automatically pings both directions on active connections. These pings are at regular intervals, and allow detection of unexpected disconnects (for example, unplugging a server). If the client detects that the server has stopped pinging, the client will close the connection, and vice versa. If there's other traffic through the connection, keep-alive pings aren't needed. A `Ping` is only sent if the interval has elapsed without a message being sent.
 
+## Upload streaming
```
Comment (mikaelm12) [inline src/SignalR/specs/HubProtocol.md]: Move this to be right below the streaming section

**Review by mikaelm12 — APPROVED** [review]
Some comments on the spec. Let's get this in

File: src/SignalR/specs/HubProtocol.md (hunk 18 above)
Comment (BrennanConroy) [reply src/SignalR/specs/HubProtocol.md]: >The Callees doesn't have to initiate this though, right?

Initiate what?

Comment (mikaelm12) [reply src/SignalR/specs/HubProtocol.md]: When you write "Sometimes Callees will want to receive streaming data from Callers"

**Comment by BrennanConroy** [comment]
Should we put the streaming from client feature behind a check for the `MinorVersion` from the server? That way we can fail with a nice clean error on the client if a newer client tries to stream to an older server.

**Comment by BrennanConroy** [comment]
Started doing some microbenchmarking with the changes:



Method |         Mean |      Error |     StdDev |      Op/s |  Gen 0 | Allocated |
---------------------- |-------------:|-----------:|-----------:|----------:|-------:|----------:|
InvocationAsync |     1.344 us |  0.0267 us |  0.0625 us | 744,175.5 | 0.0057 |     696 B |
UploadStream_One |    44.187 us |  0.1145 us |  0.0894 us |  22,630.9 |      - |    3496 B |

This shows that there is greater up-front costs for client->server streaming

Method |         Mean |      Error |     StdDev |      Op/s |  Gen 0 | Allocated |
---------------------- |-------------:|-----------:|-----------:|----------:|-------:|----------:|
UploadStream_Thousand |   318.673 us |  1.1018 us |  0.9767 us |   3,138.0 |      - |   49074 B |
Invoke_Thousand | 1,349.203 us | 21.8902 us | 20.4761 us |     741.2 | 5.8594 |  728000 B |

And lower per-item cost when comparing sending 1000 of the same messages in streaming vs invoking.

**Review by halter73 — APPROVED** [review]

File: src/SignalR/benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs [hunk 20]
```diff
@@ -163,61 +163,71 @@ public ChannelReader<int> StreamChannelReaderCount(int count)
 
                 return channel.Reader;
             }
-        }
```
Comment (davidfowl) [inline src/SignalR/benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs]: Hmm?

**Pull request merged**